tools/pythonapi/dftbplus.py

```diff
@@ -0,0 +1,274 @@
+
+        # Register supported (DFTB+) routines
+        self._setup_interface()
+
+        if self._logfile is not None:
+            output_str = ctypes.create_string_buffer(str.encode(self._logfile))
+            self._dftbpluslib.dftbp_init(self._dftb_handler, output_str)
+        else:
+            self._dftbpluslib.dftbp_init(self._dftb_handler, None)
+
+        self._dftbpluslib.dftbp_get_input_from_file(self._dftb_handler,
+                                                    input_str, self._dftb_input)
+
+        self._dftbpluslib.dftbp_process_input(self._dftb_handler,
+                                              self._dftb_input)
+
+        self._natoms = self._dftbpluslib.dftbp_get_nr_atoms(self._dftb_handler)
+
+        self._gradients = np.zeros((self._natoms, 3))
+        self._energy = np.zeros(1)
+        self._grosschg = np.zeros(self._natoms)
+
+
+    def set_geometry(self, coords, latvecs=None, relcoords=False):
+        '''Sets up the desired geometry.
+
+        Args:
+            coords (2darray):    absolute atomic positions
+                                 (in atomic units)
+            latvecs (2darray):   lattice vectors (in atomic units)
+                                 (None for non-periodic structures)
+            relcoords (2darray): relative atomic positions (in atomic units)
+        '''
+
+        self._coords = coords
+        self._natoms = len(self._coords)
+        self._latvecs = latvecs
+        self._periodic = latvecs is not None
+
+        if self._periodic:
+            self._latvecs = np.array(latvecs, dtype=float)
+            self._invlatvecs = la.inv(self._latvecs)
+            if relcoords:
+                self._relcoords = coords
+                self._coords = np.dot(self._relcoords, self._latvecs)
+            else:
+                self._coords = coords
+                self._relcoords = np.dot(self._coords, self._invlatvecs)
+        else:
+            self._latvecs = None
+            self._invlatvecs = None
+            self._relcoords = None
+
+        if self._periodic:
+            self._dftbpluslib.dftbp_set_coords_and_lattice_vecs(
+                self._dftb_handler, self._coords, self._latvecs)
+        else:
+            self._dftbpluslib.dftbp_set_coords(self._dftb_handler, self._coords)
+
+
+    def set_ext_pot(self, extpot, extpotgrad=None):
+        '''Sets up an external potential.
+
+        Args:
+            extpot (2darray):     External potential at the position of each
+                                  atom. Shape: [natom]. (in atomic units)
+            extpotgrad (2darray): Gradient of the external potential at each
+                                  atom. Shape: [natom, 3]. (in atomic units)
+                                  This parameter is optional, you can pass None
+                                  if you did not ask DFTB+ to calculate forces.
+        '''
+
+        self._extpot = extpot
```

Comment:
I think, we should not to store the queried/passed entities in `self`. It is not needed anywhere.

Reply:
You are right, storing those entities was unnecessary in many cases. I have changed this (cf. e0400dd).